tests/pipeline_with_optional_test.py

```diff
@@ -0,0 +1,140 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+from typing import NewType, Optional, Union
+
+import pytest
+
+import sciline as sl
+
+
+def test_provider_returning_optional_disallowed() -> None:
+    def make_optional() -> Optional[int]:
+        return 3
+
+    with pytest.raises(ValueError):
+        sl.Pipeline([make_optional])
+
+
+def test_provider_returning_union_disallowed() -> None:
+    def make_union() -> Union[int, float]:
+        return 3
+
+    with pytest.raises(ValueError):
+        sl.Pipeline([make_union])
+
+
+def test_parameter_type_union_or_optional_disallowed() -> None:
+    pipeline = sl.Pipeline()
+    with pytest.raises(ValueError):
+        pipeline[Union[int, float]] = 3  # type: ignore[index]
+    with pytest.raises(ValueError):
+        pipeline[Optional[int]] = 3  # type: ignore[index]
+
+
+def test_union_requirement_leads_to_UnsatisfiedRequirement() -> None:
+    def require_union(x: Union[int, float]) -> str:
+        return f'{x}'
+
+    pipeline = sl.Pipeline([require_union])
```

Comment:
Can you add a provider for `int` and/or `float` so this actually tests something?

Reply:
Good point, see update!